### PR TITLE
Native gRPC health checker implementation

### LINKED ISSUE #369: gRPC health checking against a predefined proto
Feature request.
Envoy could come up with a standard for grpc health-check eg: 
https://github.com/grpc/grpc/blob/master/doc/health-checking.md 

```
syntax = "proto3";

package grpc.health.v1;

message HealthCheckRequest {
  string service = 1;
}

message HealthCheckResponse {
  enum ServingStatus {
    UNKNOWN = 0;
    SERVING = 1;
    NOT_SERVING = 2;
  }
  ServingStatus status = 1;
}

service Health {
  rpc Check(HealthCheckRequest) returns (HealthCheckResponse);
}
```
Anyone wanting to do this would have to implement this and configure the check to be grpc.

### PR DESCRIPTION
**Pull request by baranov1ch**
Signed-off-by: Alexey Baranov <me@kotiki.cc>

*title*: *gRPC health checker*

*Description*:
This PR introduces gRPC-based health checking cluster hosts that implement [grpc.health.v1.Health](https://github.com/grpc/grpc/blob/master/src/proto/grpc/health/v1/health.proto). New `grpc_health_check` option is added to `health_checks` section of CDS `v2.Cluster` config. User can also set `service_name` which will be passed in `HealthCheckRequest.service` in `Check` RPC request payload. Cluster *MUST* be configured to support HTTP/2 to use gRPC health check, otherwise exception is thrown during configuration.

*Risk Level*: Medium

*Testing*:
Rather comprehensive (imo :)) unit testing for configuration and operation. Also it was manually tested against grpc-go based server to catch a lot of graceful shutdown edge cases (see long comments in `GrpcHealthCheckerImpl::GrpcActiveHealthCheckSession`).

*Docs Changes*:
https://github.com/envoyproxy/data-plane-api/pull/383

*Release Notes*:
Added gRPC healthcheck based on [grpc.health.v1.Health](https://github.com/grpc/grpc/blob/master/src/proto/grpc/health/v1/health.proto) service.

Fixes #369

*API Changes*:
https://github.com/envoyproxy/data-plane-api/pull/383

### CODE REVIEW

**Comment by baranov1ch** [comment]
I'm not sure how timed out `//test/integration:access_log_integration_test` relates to my changes. Plus IWOMM

File: source/common/upstream/health.proto [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+syntax = "proto3";
+
+package grpc.health.v1;
```
Comment (htuch) [inline source/common/upstream/health.proto]: I think this should be sourced from the gRPC repo as an external dependency. I'm going to be looking into adding the gRPC repos as an external dep for Envoy today, so stay tuned.

File: source/common/upstream/health_checker_impl.h [hunk 2]
```diff
@@ -417,5 +421,87 @@ class RedisHealthCheckerImpl : public HealthCheckerImplBase {
...
+      void onAboveWriteBufferHighWatermark() override {}
+      void onBelowWriteBufferLowWatermark() override {}
+
+    private:
+      GrpcActiveHealthCheckSession& parent_;
+    };
+
+    ConnectionCallbackImpl connection_callback_impl_{*this};
+    GrpcHealthCheckerImpl& parent_;
+    Http::CodecClientPtr client_;
```
Comment (htuch) [inline source/common/upstream/health_checker_impl.h]: Before going into the review, I'd like to understand some design stuff. How come we're doing all the low level gRPC codec and connection management in this HC rather than leveraging the existing gRPC client support in Envoy (https://github.com/envoyproxy/envoy/blob/master/include/envoy/grpc/async_client.h, https://github.com/envoyproxy/envoy/blob/master/source/common/grpc/async_client_impl.h)?

Comment (baranov1ch) [reply source/common/upstream/health_checker_impl.h]: grpc AsyncClientImpl was my first thought. Unfortunately async client is tied to cluster manager, router, retries, AsyncHttpClient, etc. We don't have (and don't need) all that abstractions in healthchecker code - we need to connect directly to a specific endpoint.

Other option is to refactor out connection establishment part from current Grpc::AsyncStreamImpl, leaving it as a dumb protocol decoder, but that requires careful thinking and greatly widens scope of this pretty germetic PR, which I would like to avoid and work on tech debt later. Imo, low-level protocol stuff here is pretty simple.

Comment (htuch) [reply source/common/upstream/health_checker_impl.h]: Got it, I think this is fine then, will take a deeper look in a bit.

Comment (mattklein123) [reply source/common/upstream/health_checker_impl.h]: FWIW the issue of doing it this way vs. AsyncClient came up when @craffert0 took a quick stab at implementing this a couple of months ago. I think it's definitely possible to refactor AsyncClient so it doesn't depend on cluster manager, but it's non-trivial. I'm in favor of doing it this way and adding a TODO to come back and revisit at a later time.

File: source/common/upstream/health.proto (hunk 1 above)
Comment (htuch) [reply source/common/upstream/health.proto]: See https://github.com/envoyproxy/envoy/pull/2318. When this merges, you should be able to do something like `@grpc//src/proto/grpc/health/v1:health.proto` to reference the proto there.

File: source/common/upstream/health_checker_impl.cc [hunk 3]
```diff
@@ -41,6 +43,13 @@ HealthCheckerSharedPtr HealthCheckerFactory::create(const envoy::api::v2::Health
   case envoy::api::v2::HealthCheck::HealthCheckerCase::kRedisHealthCheck:
     return std::make_shared<RedisHealthCheckerImpl>(cluster, hc_config, dispatcher, runtime, random,
                                                     Redis::ConnPool::ClientFactoryImpl::instance_);
+  case envoy::api::v2::HealthCheck::HealthCheckerCase::kGrpcHealthCheck:
+    if (!(cluster.info()->features() & Upstream::ClusterInfo::Features::HTTP2)) {
+      throw EnvoyException(fmt::format("{} cluster must support http2 for gRPC healthchecking",
```
Comment (htuch) [inline source/common/upstream/health_checker_impl.cc]: Nit: slight preference for "HTTP/2" over "http2" here.

File: source/common/upstream/health_checker_impl.cc [hunk 4]
```diff
@@ -586,5 +595,297 @@ RedisHealthCheckerImpl::HealthCheckRequest::HealthCheckRequest() {
...
+}
+
+void GrpcHealthCheckerImpl::GrpcActiveHealthCheckSession::decodeHeaders(
+    Http::HeaderMapPtr&& headers, bool end_stream) {
+  const auto http_response_status = Http::Utility::getResponseStatus(*headers);
+  if (http_response_status != enumToInt(Http::Code::OK)) {
+    // https://github.com/grpc/grpc/blob/master/doc/http-grpc-status-mapping.md requires that
+    // grpc-status be used if available.
+    if (end_stream) {
+      auto grpc_status = Grpc::Common::getGrpcStatus(*headers);
```
Comment (htuch) [inline source/common/upstream/health_checker_impl.cc]: Nit: const auto

File: source/common/upstream/health_checker_impl.cc [hunk 5]
```diff
@@ -586,5 +595,297 @@ RedisHealthCheckerImpl::HealthCheckRequest::HealthCheckRequest() {
...
+      if (grpc_status.valid()) {
+        onRpcComplete(grpc_status.value(), Grpc::Common::getGrpcMessage(*headers), true);
```
Comment (htuch) [inline source/common/upstream/health_checker_impl.cc]: Isn't gRPC always 200? If it's non-200, it seems like we shouldn't be expecting `grpc-status` at all.

File: source/common/upstream/health_checker_impl.cc [hunk 6]
```diff
@@ -586,5 +595,297 @@ RedisHealthCheckerImpl::HealthCheckRequest::HealthCheckRequest() {
...
+  }
+  if (end_stream) {
+    // This is how, for instance, grpc-go signals about missing service - HTTP/2 200 OK with
+    // 'unimplemented' gRPC status.
+    auto grpc_status = Grpc::Common::getGrpcStatus(*headers);
+    if (grpc_status.valid()) {
+      onRpcComplete(grpc_status.value(), Grpc::Common::getGrpcMessage(*headers), true);
+      return;
+    }
+    onRpcComplete(Grpc::Status::GrpcStatus::Internal, "unexpected stream end", true);
```
Comment (htuch) [inline source/common/upstream/health_checker_impl.cc]: This is a gRPC protocol violation, so maybe be a bit more explicit about that.

File: source/common/upstream/health_checker_impl.cc [hunk 7]
```diff
@@ -586,5 +595,297 @@ RedisHealthCheckerImpl::HealthCheckRequest::HealthCheckRequest() {
...
+
+  grpc::health::v1::HealthCheckRequest request;
+  if (parent_.service_name_.valid())
+    request.set_service(parent_.service_name_.value());
+
+  request_encoder_->encodeData(*Grpc::Common::serializeBody(request), true);
+}
+
+void GrpcHealthCheckerImpl::GrpcActiveHealthCheckSession::onResetStream(Http::StreamResetReason) {
+  bool expect_reset = expect_reset_;
```
Comment (htuch) [inline source/common/upstream/health_checker_impl.cc]: Nit: const bool

File: source/common/upstream/health_checker_impl.cc [hunk 8]
```diff
@@ -586,5 +595,297 @@ RedisHealthCheckerImpl::HealthCheckRequest::HealthCheckRequest() {
...
+  resetState();
+
+  if (expect_reset)
```
Comment (htuch) [inline source/common/upstream/health_checker_impl.cc]: Nit: use braces even for single line `if`.

File: source/common/upstream/health_checker_impl.cc [hunk 9]
```diff
@@ -586,5 +595,297 @@ RedisHealthCheckerImpl::HealthCheckRequest::HealthCheckRequest() {
...
+    // Force connection close on healthcheck errors and timeouts.
+    // Rationale: healthcheck error may mean cluster host is going down gracefully. Technically (
+    // for google-written gRPC server libs), it means that host sends us GOAWAY and PING, waiting
+    // PING ACK to initiate real connection close. But nghttp2 does not respond to pings received
+    // after GOAWAY, and envoy does not handle remote-initiated GOAWAYs yet
+    // (https://github.com/nghttp2/nghttp2/issues/1103).
+    // Altogether, this results in us holding the connection, preventing gRPC server graceful
+    // shutdown until server close timeout expires (typically 1 minute). This is bad. It's better to
+    // drop connection in case of errors than to mess up backend graceful shutdown. Still, all
+    // successful checks will be performed on a single HTTP/2 connection.
```
Comment (htuch) [inline source/common/upstream/health_checker_impl.cc]: How does any of this differ from other cases where we need to handle GOAWAYs and connection closes on Envoy's data or control planes?

File: source/common/upstream/health_checker_impl.cc [hunk 10]
```diff
@@ -586,5 +595,297 @@ RedisHealthCheckerImpl::HealthCheckRequest::HealthCheckRequest() {
...
+    client_->close();
+  }
+}
+
+bool GrpcHealthCheckerImpl::GrpcActiveHealthCheckSession::isHealthCheckSucceeded(
+    Grpc::Status::GrpcStatus grpc_status) const {
+  if (grpc_status != Grpc::Status::GrpcStatus::Ok)
+    return false;
```
Comment (htuch) [inline source/common/upstream/health_checker_impl.cc]: Use braces on single line `if`

File: source/common/upstream/health_checker_impl.cc [hunk 11]
```diff
@@ -586,5 +595,297 @@ RedisHealthCheckerImpl::HealthCheckRequest::HealthCheckRequest() {
...
+    handleSuccess();
+  } else {
+    handleFailure(FailureType::Active);
+  }
+
+  if (end_stream) {
+    resetState();
+  } else {
+    // resetState() will be called by onResetStream().
+    expect_reset_ = true;
```
Comment (htuch) [inline source/common/upstream/health_checker_impl.cc]: This logic is a bit confusing to me when comparing with `Grpc::AsyncClient`. Can we use the same close local/remote as done there?

File: source/common/upstream/health_checker_impl.cc [hunk 12]
```diff
@@ -586,5 +595,297 @@ RedisHealthCheckerImpl::HealthCheckRequest::HealthCheckRequest() {
...
+      break;
+    }
+  }
+  std::string grpc_status_message;
+  if (grpc_status != Grpc::Status::GrpcStatus::Ok && !grpc_message.empty()) {
+    grpc_status_message = fmt::format("{} ({})", grpc_status, grpc_message);
+  } else {
+    grpc_status_message = fmt::format("{}", grpc_status);
+  }
+  ENVOY_CONN_LOG(debug, "hc grpc_status={} service_status={} health_flags={}", *client_,
```
Comment (htuch) [inline source/common/upstream/health_checker_impl.cc]: We're doing a bunch of work above that isn't strictly needed when log level is above debug. Is it possible to condition on log level?

File: source/common/upstream/health_checker_impl.cc (hunk 5 above)
Comment (baranov1ch) [reply source/common/upstream/health_checker_impl.cc]: this code was taken from `grpc_async_client_impl.h`. I haven't managed to receive such a thing from grpc-go server, but maybe it can exist

File: source/common/upstream/health_checker_impl.cc (hunk 12 above)
Comment (baranov1ch) [reply source/common/upstream/health_checker_impl.cc]: I haven't found how to do that, but I'm quite new to the codebase. Would appreciate any help :)

File: source/common/upstream/health_checker_impl.cc (hunk 5 above)
Comment (baranov1ch) [reply source/common/upstream/health_checker_impl.cc]: This is copied from [Grpc::AsyncClentImpl](https://github.com/envoyproxy/envoy/blob/master/source/common/grpc/async_client_impl.h#L111). I wasn't able to get to this codepath from grpc-go server, but maybe it can happen with more intermediaries en route.

File: source/common/upstream/health_checker_impl.cc (hunk 9 above)
Comment (baranov1ch) [reply source/common/upstream/health_checker_impl.cc]: Other HTTP/2 cases all go through cluster manager and `Http2::ConnPool` which handles remote goaways implementing `Http::ConnectionCallbacks` by closing the client. I'd better use it here too, thanks for pointing out.

File: source/common/upstream/health_checker_impl.cc (hunk 11 above)
Comment (baranov1ch) [reply source/common/upstream/health_checker_impl.cc]: here we don't use local/remote close, logic is simpler:
- 'local' is always closed, RPC is unary, everything is sent by `onInterval()`
- `end_stream` argument is passed to `onRpcComplete` instead of using a field indicating 'remote close'.
Here local/remote close logic is not necessary, since there are no clients trying to send/receive streams of grpc messages.
- besides local/remote close, there is `http_reset_` in `Grpc::AsyncClient` which is used to determine if `Http::AsyncClient::Stream::reset()` should be called to recycle http stream, or stream was already reset by http stack. `expect_reset_ ` has nearly the same semantics as `http_reset_`. But unlike `Grpc::AsyncClient`, which maps stream reset just to internal error, health checker distinguishes between active healthcheck errors (when service is in really bad mood)  and network-ish healthcheck errors produced by stream reset.

File: source/common/upstream/health_checker_impl.cc (hunk 6 above)
Comment (baranov1ch) [reply source/common/upstream/health_checker_impl.cc]: Not sure what to do here then... there is no more suitable grpc status. I can add "protocol violation: " to error message.

Comment (htuch) [reply source/common/upstream/health_checker_impl.cc]: Yeah, just add to the error message.

File: source/common/upstream/health_checker_impl.cc (hunk 5 above)
Comment (htuch) [reply source/common/upstream/health_checker_impl.cc]: Ah, fair enough. I actually wrote that code, and suspect that it's gratuitous, but let's leave it as is for consistency.

File: source/common/upstream/health_checker_impl.cc (hunk 7 above)
Comment (htuch) [reply source/common/upstream/health_checker_impl.cc]: Actually, would also rename to `expected_reset` when snapshotting.

File: source/common/upstream/health_checker_impl.cc (hunk 11 above)
Comment (htuch) [reply source/common/upstream/health_checker_impl.cc]: Fair enough. I think it would be helpful to comment here on why we might have seen an RPC completion, without hitting the end-of-stream, and why this might lead to a following `onResetStream()` callback. Thanks.

File: source/common/upstream/health_checker_impl.cc (hunk 12 above)
Comment (htuch) [reply source/common/upstream/health_checker_impl.cc]: Just checked with @mrice32, this doesn't exist today unfortunately. I've opened https://github.com/envoyproxy/envoy/issues/2330 to track.

**Comment by baranov1ch** [comment]
oops, did rebase( Promise to to do it anymore

File: source/common/upstream/health_checker_impl.cc (hunk 11 above)
Comment (baranov1ch) [reply source/common/upstream/health_checker_impl.cc]: added some comments. PTAL if they are good enough

File: source/common/upstream/health.proto (hunk 1 above)
Comment (baranov1ch) [reply source/common/upstream/health.proto]: trying like this:

```
--- a/bazel/repositories.bzl
+++ b/bazel/repositories.bzl
@@ -410,3 +410,18 @@ def _com_github_grpc_grpc():
       name = "grpc",
       actual = "@com_github_grpc_grpc//:grpc++"
     )
+    // For grpc proto compilation to work.
+    native.bind(
+        name = "protocol_compiler",
+        actual = "@com_google_protobuf_cc//:protoc",
+    )
+
+    // For grpc proto compilation to work.
+    native.bind(
+        name = "protobuf_clib",
+        actual = "@com_google_protobuf_cc//:protobuf",
+    )
+
+    native.bind(
+      name = "grpc_health_proto",
+      actual = "@com_github_grpc_grpc//src/proto/grpc/health/v1:health_proto",
+    )


--- a/source/common/upstream/BUILD
+++ b/source/common/upstream/BUILD
@@ -89,9 +89,8 @@ envoy_cc_library(
     name = "health_checker_lib",
     srcs = ["health_checker_impl.cc"],
     hdrs = ["health_checker_impl.h"],
-    external_deps = ["envoy_health_check"],
+    external_deps = ["envoy_health_check", "grpc_health_proto"],
```
does not work with some strange errors:
```
bazel build -c dbg --verbose_failures //source/common/upstream:health_checker_lib
<...>
INFO: Analysed target //source/common/upstream:health_checker_lib (0 packages loaded).
INFO: Found 1 target...
ERROR: /private/var/tmp/_bazel_baranovich/b62ff95190308ee02bfc031330094506/external/com_github_grpc_grpc/src/proto/grpc/health/v1/BUILD:21:1: output 'external/com_github_grpc_grpc/src/proto/grpc/health/v1/grpc/src/proto/grpc/health/v1/health.pb.h' was not created
ERROR: /private/var/tmp/_bazel_baranovich/b62ff95190308ee02bfc031330094506/external/com_github_grpc_grpc/src/proto/grpc/health/v1/BUILD:21:1: output 'external/com_github_grpc_grpc/src/proto/grpc/health/v1/grpc/src/proto/grpc/health/v1/health.pb.cc' was not created
ERROR: /private/var/tmp/_bazel_baranovich/b62ff95190308ee02bfc031330094506/external/com_github_grpc_grpc/src/proto/grpc/health/v1/BUILD:21:1: not all outputs were created or valid
Target //source/common/upstream:health_checker_lib failed to build
INFO: Elapsed time: 0,506s, Critical Path: 0,30s
FAILED: Build did NOT complete successfully
```

Seems like grpc proto build rules does smth strange with proto out paths...

File: source/common/upstream/health_checker_impl.cc (hunk 9 above)
Comment (baranov1ch) [reply source/common/upstream/health_checker_impl.cc]: It got much cleaner after using `onGoAway` callback

File: source/common/upstream/health_checker_impl.cc [hunk 13]
```diff
@@ -41,6 +43,13 @@ HealthCheckerSharedPtr HealthCheckerFactory::create(const envoy::api::v2::Health
   case envoy::api::v2::HealthCheck::HealthCheckerCase::kRedisHealthCheck:
     return std::make_shared<RedisHealthCheckerImpl>(cluster, hc_config, dispatcher, runtime, random,
                                                     Redis::ConnPool::ClientFactoryImpl::instance_);
+  case envoy::api::v2::HealthCheck::HealthCheckerCase::kGrpcHealthCheck:
+    if (!(cluster.info()->features() & Upstream::ClusterInfo::Features::HTTP2)) {
+      throw EnvoyException(fmt::format("{} cluster must support HTTP/2 for gRPC healthchecking",
```
Comment (mattklein123) [inline source/common/upstream/health_checker_impl.cc]: Please have a test that does EXPECT_THROW_WITH_MESSAGE to catch this case.

File: source/common/upstream/health_checker_impl.h [hunk 14]
```diff
@@ -417,5 +421,102 @@ class RedisHealthCheckerImpl : public HealthCheckerImplBase {
...
+public:
+  GrpcHealthCheckerImpl(const Cluster& cluster, const envoy::api::v2::HealthCheck& config,
+                        Event::Dispatcher& dispatcher, Runtime::Loader& runtime,
+                        Runtime::RandomGenerator& random);
+
+private:
+  struct GrpcActiveHealthCheckSession : public ActiveHealthCheckSession,
+                                        public Http::StreamDecoder,
+                                        public Http::StreamCallbacks {
+    GrpcActiveHealthCheckSession(GrpcHealthCheckerImpl& parent, HostSharedPtr host);
```
Comment (mattklein123) [inline source/common/upstream/health_checker_impl.h]: nit: `const HostSharedPtr& host` (Please fix elsewhere in this file in old code if similar).

File: source/common/upstream/health_checker_impl.h [hunk 15]
```diff
@@ -417,5 +421,102 @@ class RedisHealthCheckerImpl : public HealthCheckerImplBase {
...
+    // e.g. remote reset. In this case healthcheck status has already been reported, only state
+    // cleanup is required.
+    bool expect_reset_ = false;
+  };
+
+  virtual Http::CodecClientPtr createCodecClient(Upstream::Host::CreateConnectionData& data) PURE;
+
+  // HealthCheckerImplBase
+  ActiveHealthCheckSessionPtr makeSession(HostSharedPtr host) override {
+    return ActiveHealthCheckSessionPtr{new GrpcActiveHealthCheckSession(*this, host)};
```
Comment (mattklein123) [inline source/common/upstream/health_checker_impl.h]: nit: std::make_unique (please fix elsewhere in this file for similar code, since we are using C++14 now).

File: source/common/upstream/health_checker_impl.cc [hunk 16]
```diff
@@ -586,5 +595,265 @@ RedisHealthCheckerImpl::HealthCheckRequest::HealthCheckRequest() {
...
+      break;
+    }
+  }
+  std::string grpc_status_message;
+  if (grpc_status != Grpc::Status::GrpcStatus::Ok && !grpc_message.empty()) {
+    grpc_status_message = fmt::format("{} ({})", grpc_status, grpc_message);
+  } else {
+    grpc_status_message = fmt::format("{}", grpc_status);
+  }
+  ENVOY_CONN_LOG(error, "hc grpc_status={} service_status={} health_flags={}", *client_,
```
Comment (mattklein123) [inline source/common/upstream/health_checker_impl.cc]: I think this should actually be debug level. (Since upstream can error and we don't want to spew logs by default).

File: source/common/upstream/health_checker_impl.cc [hunk 17]
```diff
@@ -586,5 +595,265 @@ RedisHealthCheckerImpl::HealthCheckRequest::HealthCheckRequest() {
...
+                  "gRPC protocol violation: unexpected stream end", true);
+    return;
+  }
+
+  // We should end up with only one frame here.
+  std::vector<Grpc::Frame> decoded_frames;
+  if (!decoder_.decode(data, decoded_frames)) {
+    onRpcComplete(Grpc::Status::GrpcStatus::Internal, "gRPC wire protocol decode error", false);
+  }
+  for (auto& frame : decoded_frames) {
```
Comment (mattklein123) [inline source/common/upstream/health_checker_impl.cc]: Do you have a test with the response split across two decodeData calls?

File: source/common/upstream/health_checker_impl.cc [hunk 18]
```diff
@@ -586,5 +595,265 @@ RedisHealthCheckerImpl::HealthCheckRequest::HealthCheckRequest() {
...
+
+GrpcHealthCheckerImpl::GrpcActiveHealthCheckSession::~GrpcActiveHealthCheckSession() {
+  if (client_) {
+    // If there is an active request it will get reset, so make sure we ignore the reset.
+    expect_reset_ = true;
+    client_->close();
+  }
+}
+
+void GrpcHealthCheckerImpl::GrpcActiveHealthCheckSession::decodeHeaders(
```
Comment (mattklein123) [inline source/common/upstream/health_checker_impl.cc]: I haven't looked at the coverage results yet in CI, but please check and make sure you have coverage on all error branches in the code you added.

File: source/common/upstream/health.proto (hunk 1 above)
Comment (htuch) [reply source/common/upstream/health.proto]: It's hard to know what's going on with just that error message. The diff you provide looks correct. Can you do `bazel build -s` and attach the `protoc` invocation on `health.proto`? This might give more of a clue about what is going on.

File: source/common/upstream/health_checker_impl.cc (hunk 11 above)
Comment (htuch) [reply source/common/upstream/health_checker_impl.cc]: Yeah, these are good, thanks. Maybe also explain how we ever get into the situation where we are in this branch, i.e. who called `onRpcComplete` with `!end_stream` as well.

File: source/common/upstream/health.proto (hunk 1 above)
Comment (baranov1ch) [reply source/common/upstream/health.proto]:
```
SUBCOMMAND: # @com_github_grpc_grpc//src/proto/grpc/health/v1:_health_proto_codegen [action 'SkylarkAction external/com_github_grpc_grpc/src/proto/grpc/health/v1/grpc/src/proto/grpc/health/v1/health.pb.h']
(cd /private/var/tmp/_bazel_baranovich/b62ff95190308ee02bfc031330094506/execroot/envoy && \
  exec env - \
  bazel-out/host/bin/external/com_google_protobuf_cc/protoc '--cpp_out=:bazel-out/darwin-dbg/genfiles' '-Iexternal/com_github_grpc_grpc/src/proto/grpc/health/v1/health.proto=external/com_github_grpc_grpc/src/proto/grpc/health/v1/health.proto' external/com_github_grpc_grpc/src/proto/grpc/health/v1/health.proto)
```

File: source/common/upstream/health_checker_impl.cc (hunk 17 above)
Comment (baranov1ch) [reply source/common/upstream/health_checker_impl.cc]: added the test

File: source/common/upstream/health_checker_impl.cc (hunk 18 above)
Comment (baranov1ch) [reply source/common/upstream/health_checker_impl.cc]: checked, added more coverage

File: source/common/upstream/health_checker_impl.cc (hunk 13 above)
Comment (baranov1ch) [reply source/common/upstream/health_checker_impl.cc]: done

File: source/common/upstream/health_checker_impl.cc (hunk 11 above)
Comment (baranov1ch) [reply source/common/upstream/health_checker_impl.cc]: done

File: source/common/upstream/health_checker_impl.cc (hunk 16 above)
Comment (baranov1ch) [reply source/common/upstream/health_checker_impl.cc]: fixed. Just forgot to return back after testing

File: source/common/upstream/health_checker_impl.h (hunk 14 above)
Comment (baranov1ch) [reply source/common/upstream/health_checker_impl.h]: done

File: source/common/upstream/health_checker_impl.h (hunk 15 above)
Comment (baranov1ch) [reply source/common/upstream/health_checker_impl.h]: done

File: test/common/upstream/health_checker_impl_test.cc [hunk 19]
```diff
@@ -1452,6 +1494,860 @@ TEST_F(RedisHealthCheckerImplTest, AllDontReuseConnection) {
...
+  void respondServiceStatus(size_t index,
+                            grpc::health::v1::HealthCheckResponse::ServingStatus status) {
+    respondResponseSpec(index,
+                        ResponseSpec{{{":status", "200"}, {"content-type", "application/grpc"}},
+                                     {ResponseSpec::validChunk(status)},
+                                     {{"grpc-status", "0"}}});
+  }
+
+  void respondResponseSpec(size_t index, ResponseSpec&& spec) {
+    bool trailers_empty = spec.trailers.size() == 0U;
```
Comment (htuch) [inline test/common/upstream/health_checker_impl_test.cc]: Nit: can you constify the local tmps where they are immutable everywhere in the tests? Thanks.

File: test/common/upstream/health_checker_impl_test.cc [hunk 20]
```diff
@@ -1452,6 +1494,860 @@ TEST_F(RedisHealthCheckerImplTest, AllDontReuseConnection) {
...
+  EXPECT_CALL(runtime_.snapshot_, getInteger("health_check.max_interval", _));
+  EXPECT_CALL(runtime_.snapshot_, getInteger("health_check.min_interval", _))
+      .WillOnce(Return(45000));
+  EXPECT_CALL(*test_sessions_[0]->interval_timer_, enableTimer(std::chrono::milliseconds(45000)));
+  EXPECT_CALL(*test_sessions_[0]->timeout_timer_, disableTimer());
+  respondServiceStatus(0, grpc::health::v1::HealthCheckResponse::SERVING);
+  EXPECT_TRUE(cluster_->prioritySet().getMockHostSet(0)->hosts_[0]->healthy());
+}
+
+TEST_F(GrpcHealthCheckerImplTest, SuccessResponseSplitBetweenChunks) {
```
Comment (htuch) [inline test/common/upstream/health_checker_impl_test.cc]: Generally prefer a single line `//` comment above each test explaining in plain language what the test is trying to accomplish. E.g. `// Validate that a regular response split across chunks is handled.`

File: test/common/upstream/health_checker_impl_test.cc [hunk 21]
```diff
@@ -1452,6 +1494,860 @@ TEST_F(RedisHealthCheckerImplTest, AllDontReuseConnection) {
...
+  EXPECT_CALL(*this, onHostStatus(_, false)).Times(2);
+
+  cluster_->prioritySet().getMockHostSet(0)->hosts_ = {
+      makeTestHost(cluster_->info_, "tcp://127.0.0.1:80")};
+  expectSessionCreate();
+  expectStreamCreate(0);
+  EXPECT_CALL(*test_sessions_[0]->timeout_timer_, enableTimer(_));
+  health_checker_->start();
+
+  EXPECT_CALL(*test_sessions_[0]->interval_timer_, enableTimer(_));
```
Comment (htuch) [inline test/common/upstream/health_checker_impl_test.cc]: Can this be factored out into some expect function for `healthy()`? Seems very verbose to have this enable/disable timer stuff everywhere, which makes it a bit harder to read.

File: test/common/upstream/health_checker_impl_test.cc [hunk 22]
```diff
@@ -1452,6 +1494,860 @@ TEST_F(RedisHealthCheckerImplTest, AllDontReuseConnection) {
...
+    }));
+
+TEST_P(BadResponseGrpcHealthCheckerImplTest, GrpcBadResponse) {
+  setupHC();
+  cluster_->prioritySet().getMockHostSet(0)->hosts_ = {
+      makeTestHost(cluster_->info_, "tcp://127.0.0.1:80")};
+  expectSessionCreate();
+  expectStreamCreate(0);
+  EXPECT_CALL(*test_sessions_[0]->timeout_timer_, enableTimer(_));
+  health_checker_->start();
```
Comment (htuch) [inline test/common/upstream/health_checker_impl_test.cc]: I think there is also some common setup code that could be factored out.

Comment (baranov1ch) [reply test/common/upstream/health_checker_impl_test.cc]: done

File: test/common/upstream/health_checker_impl_test.cc (hunk 21 above)
Comment (baranov1ch) [reply test/common/upstream/health_checker_impl_test.cc]: done.

File: test/common/upstream/health_checker_impl_test.cc (hunk 20 above)
Comment (baranov1ch) [reply test/common/upstream/health_checker_impl_test.cc]: done

File: test/common/upstream/health_checker_impl_test.cc (hunk 19 above)
Comment (baranov1ch) [reply test/common/upstream/health_checker_impl_test.cc]: done

**Comment by htuch** [comment]
Thanks for the test cleanups, this is rad. If we can get the health.proto build sorted this LGTM.

**Comment by htuch** [comment]
@baranov1ch I looked into the health.proto build issue, fix is at https://gist.github.com/htuch/ff95a662d3c5684eb98d005f6bbbdb8e.

**Comment by baranov1ch** [comment]
@htuch thanks, it worked. I wish I could understand why :)
